### PR TITLE
Support `map_keys` & `map_values` for MAP type

### LINKED ISSUE #12147: Support `map_keys` function for Map type
### Is your feature request related to a problem or challenge?

Description | Return a list of all keys in the map.
-- | --
Example | map_keys(map([100, 5], [42,43]))
Result | [100, 5]

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

https://docs.databricks.com/en/sql/language-manual/functions/map_keys.html#map_keys-function
https://duckdb.org/docs/sql/functions/map.html#map_keysmap

### LINKED ISSUE #12148: Support `map_values` for MAP type
### Is your feature request related to a problem or challenge?

Description | Return a list of all values in the map.
-- | --
Example | map_values(map([100, 5], [42, 43]))
Result | [42, 43]

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

https://duckdb.org/docs/sql/functions/map.html#map_valuesmap
https://docs.databricks.com/en/sql/language-manual/functions/map_values.html

### PR DESCRIPTION
**Pull request by dharanad**
## Which issue does this PR close?

Closes #12147 & #12148

## What changes are included in this PR?
* Added `map_keys` & `map_values` scalar function. Ref : https://duckdb.org/docs/sql/functions/map.html#map_valuesmap

### CODE REVIEW

**Comment by dharanad** [comment]
Encounter a bug. Fixed it by changing field name to `item`. But why ?
```
Internal error: Failed due to a difference in schemas, original schema: DFSchema { inner: Schema { fields: [Field { name: "map_keys(map(make_array(Int64(1),Int64(2),Int64(3)),make_array(Int64(1),Int64(2),Int64(3))))", data_type: List(Field { name: "keys", data_type: Int64, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }, field_qualifiers: [None], functional_dependencies: FunctionalDependencies { deps: [] } }, new schema: DFSchema { inner: Schema { fields: [Field { name: "map_keys(map(make_array(Int64(1),Int64(2),Int64(3)),make_array(Int64(1),Int64(2),Int64(3))))", data_type: List(Field { name: "item", data_type: Int64, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }), nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }, field_qualifiers: [None], functional_dependencies: FunctionalDependencies { deps: [] } }.
```

**Comment by Weijun-H** [comment]
> Encounter a bug. Fixed it by changing field name to `item`. But why ?
> 
> ```
> Internal error: Failed due to a difference in schemas, original schema: DFSchema { inner: Schema { fields: [Field { name: "map_keys(map(make_array(Int64(1),Int64(2),Int64(3)),make_array(Int64(1),Int64(2),Int64(3))))", data_type: List(Field { name: "keys", data_type: Int64, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }, field_qualifiers: [None], functional_dependencies: FunctionalDependencies { deps: [] } }, new schema: DFSchema { inner: Schema { fields: [Field { name: "map_keys(map(make_array(Int64(1),Int64(2),Int64(3)),make_array(Int64(1),Int64(2),Int64(3))))", data_type: List(Field { name: "item", data_type: Int64, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }), nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }, field_qualifiers: [None], functional_dependencies: FunctionalDependencies { deps: [] } }.
> ```

I think it is caused by the pre-defined field name in `make_array`.

https://github.com/apache/datafusion/blob/7e9ea3ad59071d56093c197c5ecd5c50021deb94/datafusion/functions-nested/src/map.rs#L33-L39

File: datafusion/sqllogictest/test_files/map.slt [hunk 1]
```diff
@@ -568,8 +568,49 @@ select map_extract(column1, 1), map_extract(column1, 5), map_extract(column1, 7)
...
+----
+[[1, 2, 3]]
+[[4]]
+[[1, 2]]
+
+query ?
+SELECT map_keys(Map{});
+----
+[]
+
```
[the remark this answers is not in the document]
Comment (Weijun-H) [reply datafusion/sqllogictest/test_files/map.slt:610]: How about adding some tests for columnar?

[the remark this answers is not in the document]
Comment (dharanad) [reply datafusion/sqllogictest/test_files/map.slt:610]: Great idea. Adding the suggested test case

**Review by Weijun-H — APPROVED** [review]
LGTM! Thanks @dharanad

File: datafusion/functions-nested/src/utils.rs [hunk 2]
```diff
@@ -253,6 +254,21 @@ pub(crate) fn compute_array_dims(
...
 }
 
+pub(crate) fn get_map_entry_field(data_type: &DataType) -> Result<&Fields> {
+    match data_type {
+        DataType::Map(field, _) => {
+            let field_data_type = field.data_type();
+            match field_data_type {
+                DataType::Struct(fields) => Ok(fields),
+                _ => {
+                    _internal_err!("Expected a Struct type, got {:?}", field_data_type)
```
Comment (jayzhan211) [inline datafusion/functions-nested/src/utils.rs:264]: nit: `internal_err`

File: datafusion/functions-nested/src/lib.rs [hunk 3]
```diff
@@ -43,6 +43,7 @@ pub mod length;
 pub mod make_array;
 pub mod map;
 pub mod map_extract;
+mod map_keys;
```
Comment (Blizzara) [inline datafusion/functions-nested/src/lib.rs:46]: pub mod?

**Comment by Weijun-H** [comment]
@dharanad could you add some doc for this function?

**Comment by dharanad** [comment]
Sorry for the mess, i thought since the change was similar for both `map_keys` && `map_values` why not club it.

File: docs/source/user-guide/sql/scalar_functions.md [hunk 4]
```diff
@@ -3729,6 +3731,56 @@ SELECT map_extract(MAP {'a': 1, 'b': NULL, 'c': 3}, 'a');
...
+  Can be a constant, column, or function, and any combination of map operators.
+
+#### Example
+
+```
+SELECT map_values(MAP {'a': 1, 'b': NULL, 'c': 3});
+----
+[1, , 3]
+
+select map_keys(map([100, 5], [42,43]));
```
Comment (Weijun-H) [inline docs/source/user-guide/sql/scalar_functions.md:3779]:
```suggestion
select map_values(map([100, 5], [42,43]));
```

File: datafusion/functions-nested/src/map_keys.rs [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+#[derive(Debug)]
+pub(crate) struct MapKeysFunc {
+    signature: Signature,
+}
+
+impl MapKeysFunc {
+    pub fn new() -> Self {
+        Self {
+            signature: Signature::user_defined(Volatility::Immutable),
```
Comment (jayzhan211) [inline datafusion/functions-nested/src/map_keys.rs:44]: This seems like Signature::Any(1)?

Comment (Weijun-H) [reply datafusion/functions-nested/src/map_keys.rs:44]: `Signature:exact(TypeSignature::ArraySignature(ArrayFunctionSignature::MapArray))` is better

Comment (dharanad) [reply datafusion/functions-nested/src/map_keys.rs:44]: To @Weijun-H suggestion looks perfect. @jayzhan211  Do you see any issues ?

**Comment by Weijun-H** [comment]
After fixing CI, it is ready to go

**Comment by dharanad** [comment]
> After fixing CI, it is ready to go


@Weijun-H Any idea, how do i fix the docs formatting issue

**Comment by Weijun-H** [comment]
> > After fixing CI, it is ready to go
> 
> @Weijun-H Any idea, how do i fix the docs formatting issue

I could check here https://datafusion.apache.org/contributor-guide/howtos.html#how-to-format-md-document

File: datafusion/functions-nested/src/map_values.rs [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    fn invoke(&self, args: &[ColumnarValue]) -> datafusion_common::Result<ColumnarValue> {
+        make_scalar_function(map_values_inner)(args)
+    }
+}
+
+fn map_values_inner(args: &[ArrayRef]) -> Result<ArrayRef> {
+    if args.len() != 1 {
+        return exec_err!("map_keys expects single argument");
```
Comment (Weijun-H) [inline datafusion/functions-nested/src/map_values.rs:88]:
```suggestion
        return exec_err!("map_values expects single argument");
```

**Comment by Weijun-H** [comment]
ship it!

**Comment by Weijun-H** [comment]
Thanks @dharanad  contribution, and @Blizzara and @jayzhan211 for reviewing.

**Comment by dharanad** [comment]
Thank You @Weijun-H @jayzhan211 @Blizzara

**Pull request merged**